### PR TITLE
Make admin event listeners general

### PR DESCRIPTION
**Pull request by CheAlex**

### CODE REVIEW

File: EventListener/MenuItemSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+     * @param OnFlushEventArgs $event
+     */
+    public function onFlush(OnFlushEventArgs $event)
+    {
+        $uow = $event->getEntityManager()->getUnitOfWork();
+
+        $isNeedRecoverTree = false;
+
+        foreach ($uow->getScheduledEntityInsertions() as $entity) {
+            if ($entity instanceof BaseMenuItem) {
```
Comment (aivus) [inline EventListener/MenuItemSubscriber.php]: Try to use "Entity Listener"

You should add tag for this listener:

```
- { name: doctrine.orm.entity_listener, entity: Grossum\MenuBundle\Entity\BaseMenuItem, event: onFlush }
```

And use

```
public function onFlush(BaseMenuItem $menuItem, OnFlushEventArgs $event)
```

File: Admin/MenuItemAdmin.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    public function setMenuItemManager(BaseMenuItemManager $menuItemManager)
+    {
+        $this->menuItemManager = $menuItemManager;
+    }
+
+    /**
+     * @param MenuManager $menuMaster
+     */
+    public function setMenuMaster(MenuManager $menuMaster)
```
Comment (aivus) [inline Admin/MenuItemAdmin.php]: I'm not sure about naming. What is it mean `master` here?

File: EventListener/MenuItemListener.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+        $this->persistedMenuItems = [];
+        $this->updatedMenuItems   = [];
+        $this->removedMenuItems   = [];
+    }
+
+    protected function recoverTree()
+    {
+        $this->getMenuItemManager()->getRepository()->recover();
```
Comment (aivus) [inline EventListener/MenuItemListener.php]: Could you explain, why do we need `recover`?

File: Form/Type/MenuEntityClassType.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+use Grossum\MenuBundle\Manager\MenuManager;
+
+class MenuEntityClassType extends AbstractType
+{
+    const NAME = 'grossum_menu_entity_class';
+    /**
```
Comment (aivus) [inline Form/Type/MenuEntityClassType.php:15]: New line missed

File: Validator/MenuItemSubjectSetted.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @Annotation
+ * @Target({"CLASS"})
+ */
+class MenuItemSubjectSetted extends Constraint
+{
+    public $message = 'The menu item "subject" must be the one';
```
Comment (aivus) [inline Validator/MenuItemSubjectSetted.php]: This message should be changed. It's not clear.

File: Validator/MenuItemSubjectSetted.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Grossum\MenuBundle\Validator;
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @Annotation
+ * @Target({"CLASS"})
+ */
+class MenuItemSubjectSetted extends Constraint
```
Comment (aivus) [inline Validator/MenuItemSubjectSetted.php]: `Set` is irregular verb.
We can't say `setted`

File: Validator/MenuItemSubjectSettedValidator.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace Grossum\MenuBundle\Validator;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+use Grossum\MenuBundle\Entity\BaseMenuItem;
+
+class MenuItemSubjectSettedValidator extends ConstraintValidator
```
Comment (aivus) [inline Validator/MenuItemSubjectSettedValidator.php]: `Set` is irregular verb.
We can't say `setted`

File: Validator/MenuItemSubjectSettedValidator.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * @param BaseMenuItem $menuItem
+     * @param Constraint $constraint
+     */
+    public function validate($menuItem, Constraint $constraint)
+    {
+        /* @var $menuItem BaseMenuItem */
+        /* @var $constraint MenuItemSubjectSetted */
```
Comment (aivus) [inline Validator/MenuItemSubjectSettedValidator.php]: Set this typehint in function phpdoc

File: Validator/MenuItemSubjectSettedValidator.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+        if (!($menuItem->getUrl() xor $menuItem->getEntityClass())) {
```
Comment (aivus) [inline Validator/MenuItemSubjectSettedValidator.php]: Move this condition to separated variables with clear name.

File: Admin/MenuItemAdmin.php (hunk 2 above)
Comment (CheAlex) [inline Admin/MenuItemAdmin.php]: Some time ago there were fields with types Grossum\MenuBundle\Entity\EntityManager\BaseMenuManager and Grossum\MenuBundle\Manager\MenuManager...
Renamed.

File: EventListener/MenuItemListener.php (hunk 3 above)
Comment (CheAlex) [inline EventListener/MenuItemListener.php]: It remained from the previous method of saving tree (when we set partial tree parameters). Now we set all tree parameters manually (taking values from js). Worked without this call. Removed.

File: Controller/MenuItemAdminController.php [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * @param Request $request
+     * @return Response
+     * @throws \Twig_Error_Runtime
+     */
+    public function treeAction(Request $request)
+    {
+        $menuId = $request->get($this->admin->getParent()->getIdParameter());
+
+        /* @var $menu BaseMenu */
```
Comment (aivus) [inline Controller/MenuItemAdminController.php]: https://phpdoc.org/docs/latest/references/phpdoc/tags/var.html

File: Controller/MenuItemAdminController.php [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+     * @param Request $request
+     * @return JsonResponse
+     */
+    public function saveTreeAction(Request $request)
+    {
+        $tree = $request->request->get('tree');
+
+        $menuId = $request->get($this->admin->getParent()->getIdParameter());
+
+        /* @var $menu BaseMenu */
```
Comment (aivus) [inline Controller/MenuItemAdminController.php]: https://phpdoc.org/docs/latest/references/phpdoc/tags/var.html

File: Controller/MenuItemAdminController.php [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        $this->admin->setSubject($subject);
+
+        $form = $this->admin->getForm();
+        $form->setData($subject);
+        $form->handleRequest($request);
+
+        $twig = $this->get('twig');
+
+        /* @var $extension FormExtension */
+        $extension =$twig->getExtension('form');
```
Comment (aivus) [inline Controller/MenuItemAdminController.php]: Missed space

File: Controller/MenuItemAdminController.php [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+        $this->admin->setSubject($subject);
+
+        $form = $this->admin->getForm();
+        $form->setData($subject);
+        $form->handleRequest($request);
+
+        $twig = $this->get('twig');
+
+        /* @var $extension FormExtension */
```
Comment (aivus) [inline Controller/MenuItemAdminController.php]: https://phpdoc.org/docs/latest/references/phpdoc/tags/var.html

File: Entity/EntityManager/BaseMenuItemManager.php [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        return $this->repository;
+    }
+
+    /**
+     * @param BaseMenu $menu
+     * @return BaseMenuItem
+     */
+    public function createRootMenuItemForMenu(BaseMenu $menu)
+    {
+        /* @var $rootMenuItem BaseMenuItem */
```
Comment (aivus) [inline Entity/EntityManager/BaseMenuItemManager.php]: https://phpdoc.org/docs/latest/references/phpdoc/tags/var.html

File: Form/EventListener/AddEntityIdentifierFieldSubscriber.php [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /**
+     * @param FormEvent $event
+     */
+    public function onPreSetData(FormEvent $event)
+    {
+        $form = $event->getForm();
+
+        /* @var $menuItem BaseMenuItem */
```
Comment (aivus) [inline Form/EventListener/AddEntityIdentifierFieldSubscriber.php]: https://phpdoc.org/docs/latest/references/phpdoc/tags/var.html

File: Validator/MenuItemSubjectSet.php [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @Annotation
+ * @Target({"CLASS"})
+ */
+class MenuItemSubjectSet extends Constraint
+{
+    public $message = 'In menu item must be set "url" or "entity class"';
```
Comment (aivus) [inline Validator/MenuItemSubjectSet.php]: `Menu item should have url or selected entity only`

**Comment by CheAlex** [comment]
This pull request depends on this https://github.com/GrossumUA/MenuBundle/pull/6 request

**Pull request merged**